File: JavaScriptPractice/Leetcode214.js
var KMP = function(s)
{
    let fast=1
    let slow=0
    let ans=Array.from({length: s.length}, () => 0)
    while (fast<s.length) {
        if (s[fast]==s[slow]) {
            slow++
            ans[fast]=slow
            fast++
        }
        else 
        {
            if (slow!=0) {
                slow = ans[slow-1]
            }
            else 
            {
                ans[fast]=0
                fast++
            }
        }
    }
    return ans
}
var shortestPalindrome = function(s) {
    let reverseS = s;
    reverseS = reverseS.split("").reverse().join("")
    let mergeS = s+" "+reverseS
    let KMPList = KMP(mergeS)

    return reverseS.slice(0,reverseS.length -KMPList[KMPList.length-1])+s;
    
};

File: CSharpPractice/LeetcodePractice/Leetcode214.cs
using System.Numerics;

public class Solution214 {
    private int[] KMP(string s)
    {
        int[] ans = new int[s.Length];
        int slow = 0;
        int fast = 1;
        while (fast<s.Length)
        {
            if (s[fast] ==s[slow])
            {
                slow++;
                ans[fast]=slow;
                fast++;
            }
            else
            {
                if (slow!=0)
                {
                    slow = ans[slow-1];
                }
                else
                {
                    ans[fast] = 0;
                    fast++;
                }
            }
        }
        return ans;
    }
    public string ShortestPalindrome(string s) {
        string reverseS = s;
        reverseS = new string(reverseS.Reverse().ToArray());
        string mergeS = s+" " +reverseS;
        int[] KMPList = KMP(s);

        return reverseS.Substring(0,reverseS.Length-KMPList[KMPList.Length-1]) +s;
        
    }
}